### PR TITLE
Sed 3347 implement new yaml plan editor or allow to load a yaml plan in the UI plan editor

### PR DESCRIPTION
**Pull request by iegorov777**

### CODE REVIEW

File: step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/LineNumberByJsonPointerResolver.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+    private static final Pattern specialChars = Pattern.compile("[\\s~!@#$%^&*()_+\\-=\\[\\]{};':\"\\\\|,.<>/?]+");
+
+    public LineNumberByJsonPointerResolver() {
+    }
+
+    public List<JsonPointerSourceLine> findLineNumbers(List<String> jsonPointers, String source) {
+        List<JsonPointerSourceLine> result = new ArrayList<>();
+
+        // TODO: maybe optimize and avoid creation of new ObjectMapper and factory for each source
```
Comment (david-stephan) [inline step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/LineNumberByJsonPointerResolver.java:53]: @iegorov777 yes, I don't think the creation of com.jayway.jsonpath.Configuration has to be done for each invocation of findLineNumbers. If it's thread safe a single instance should be enough

File: step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            PlanCompilerException planCompilerException = new PlanCompilerException();
+
+            // if the reason is validation exception (json schema) we try to resolve the source line
+            if (e.getCause() != null && e.getCause() instanceof ValidationException) {
+                ValidationException detailedValidationException = (ValidationException) e.getCause();
+                List<ValidationException> causingExceptions = detailedValidationException.getCausingExceptions();
+                if (causingExceptions != null && !causingExceptions.isEmpty()) {
+                    // sometimes the validator provides the causing exceptions, in this case we can collect several error messages with source lines
+                    List<LineNumberByJsonPointerResolver.JsonPointerSourceLine> allPointerSourceLines =
+                            new LineNumberByJsonPointerResolver().findLineNumbers(
```
Comment (david-stephan) [inline step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java:65]: Do we need a new instance of LineNumberByJsonPointerResolver each time?

File: step-controller/step-controller-server/src/main/java/step/core/plans/PlanServices.java [hunk 3]
```diff
@@ -88,19 +88,38 @@ public void init() throws Exception {
...
+		}
+		PlanType<Plan> planType = getPlanTypeNotNull(entity.getClass());
 		planType.onBeforeSave(entity);
 		return super.beforeSave(entity);
 	}
 
+	protected PlanType<Plan> getPlanTypeNotNull(String planTypeName) {
+		PlanType<Plan> planType = planTypeRegistry.getPlanType(planTypeName);
+		if (planType == null) {
+			throw new ControllerServiceException(400, "Plan type is not resolved by name " + planTypeName);
```
Comment (david-stephan) [inline step-controller/step-controller-server/src/main/java/step/core/plans/PlanServices.java:110]: @iegorov777 since this is a controller exception (visible by the user) I would rephrase it to something like "Plan type {} is not supported."

File: step-controller/step-controller-server/src/main/java/step/core/plans/PlanServices.java [hunk 4]
```diff
@@ -88,19 +88,38 @@ public void init() throws Exception {
...
 		Plan plan = planType.newPlan(template);
 		getObjectEnricher().accept(plan);
 		return plan;
 	}
 
 	@Override
 	protected Plan beforeSave(Plan entity) {
-		PlanType<Plan> planType = (PlanType<Plan>) planTypeRegistry.getPlanType(entity.getClass());
+		if (entity == null) {
+			throw new ControllerServiceException(400, "Plan is undefined");
```
Comment (david-stephan) [inline step-controller/step-controller-server/src/main/java/step/core/plans/PlanServices.java:100]: I would write something more precise, user friendly "The submitted plan to be saved is empty or null"

File: step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanType.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Override
+    public PlanCompiler<YamlEditorPlan> getPlanCompiler() {
+        return compiler;
+    }
+
+    @Override
+    public YamlEditorPlan newPlan(String template) throws Exception {
+        YamlEditorPlan plainTextPlan = new YamlEditorPlan();
```
Comment (david-stephan) [inline step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanType.java:44]: variable should be renamed to yamlPlan or yamlEditorPlan

File: step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            compilationError.setMessage(e.getMessage());
+            planCompilerException.addError(compilationError);
+            throw planCompilerException;
+        }
+
+        AbstractArtefact root = plan.getRoot();
+        AbstractArtefact parsedRoot = parsedPlan.getRoot();
+
+        // Make sure to keep the same id and attributes for root element
+        if (root != null && root.getClass() == parsedRoot.getClass()) {
```
Comment (david-stephan) [inline step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java:115]: @iegorov777  Can you explain why this is required? Also why is it not required if the class of the root change?

**Review by david-stephan — CHANGES_REQUESTED** [review]

File: step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/LineNumberByJsonPointerResolver.java (hunk 1 above)
Comment (iegorov777) [reply step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/LineNumberByJsonPointerResolver.java]: @david-stephan 
Already fixed now

File: step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java (hunk 2 above)
Comment (iegorov777) [reply step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java]: @david-stephan , already created once in constructor

File: step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java (hunk 6 above)
Comment (iegorov777) [reply step-plans/step-plans-yaml-parser/src/main/java/step/plans/parser/yaml/editor/YamlEditorPlanTypeCompiler.java:115]: @david-stephan 
I fact I don't know. This is copied from `PlainTextPlanCompiler` in Step Enterprise.

**Review by david-stephan — DISMISSED** [review]
